pkg/storage/factory.go

```diff
@@ -269,33 +273,46 @@ func NewIndexClient(name string, cfg Config, schemaCfg config.SchemaConfig, limi
 	case config.StorageTypeGrpc:
 		return grpc.NewStorageClient(cfg.GrpcConfig, schemaCfg)
 	case config.BoltDBShipperType:
-		if boltDBIndexClientWithShipper != nil {
-			return boltDBIndexClientWithShipper, nil
-		}
-
 		if shouldUseIndexGatewayClient(cfg.BoltDBShipperConfig.Config) {
+			if indexGatewayClient != nil {
+				return indexGatewayClient, nil
+			}
+
 			gateway, err := gatewayclient.NewGatewayClient(cfg.BoltDBShipperConfig.IndexGatewayClientConfig, registerer, util_log.Logger)
 			if err != nil {
 				return nil, err
 			}
 
-			boltDBIndexClientWithShipper = gateway
+			indexGatewayClient = gateway
 			return gateway, nil
 		}
 
-		objectClient, err := NewObjectClient(cfg.BoltDBShipperConfig.SharedStoreType, cfg, cm)
+		if client, ok := boltdbIndexClientsWithShipper[periodCfg.From]; ok {
+			return client, nil
+		}
+
+		objectType := periodCfg.ObjectType
+		if cfg.BoltDBShipperConfig.SharedStoreType != "" {
+			objectType = cfg.BoltDBShipperConfig.SharedStoreType
+		}
+
+		objectClient, err := NewObjectClient(objectType, cfg, cm)
 		if err != nil {
 			return nil, err
 		}
 
-		tableRanges := getIndexStoreTableRanges(config.BoltDBShipperType, schemaCfg.Configs)
-
-		boltDBIndexClientWithShipper, err = shipper.NewShipper(cfg.BoltDBShipperConfig, objectClient, limits,
-			ownsTenantFn, tableRanges, registerer)
+		tableRange := getTableRange()
+		instanceName := fmt.Sprintf("%s_%d", periodCfg.ObjectType, tableRange.Start)
+		shipper, err := shipper.NewShipper(instanceName, cfg.BoltDBShipperConfig, objectClient, limits,
```

Comment:
I think you can use `instanceName` as a static label in registerer to get rid of global metrics instances.
It would also be good to pass a logger to `instanceName` in it to be able to filter out logs by schema.

Reply:
we are already adding a component label here, that should suffice I think.
https://github.com/grafana/loki/blob/6d527e78e9474a8d5e59a5d3de989be17dfc5087/pkg/storage/store.go#L219